topics/RSA-Keys-Generation.md

```diff
@@ -0,0 +1,240 @@
+The `-p` option specifies plain hexdump without line number or ASCII character columns.
+The `-r` option reverses the operation, converting hex back to binary.
+* `base64`: Encodes the binary output from the previous step into Base64 format.
+* `tr +/ -_`: Translates the + and / characters in the Base64 output to - and _, respectively. This is a common modification for URL-safe Base64 encoding.
+* `tr -d "=\n"`: Removes any equal signs (=) and newline characters from the final Base64-encoded string.
+</tab>
+</tabs>
+
+The output of the above command is:
+
+```Shell
+$ echo "b5:f2:5a:2e:bc:d7:20:b5:20:d5:4d:cd:d4:a5:
+    7c:c8:9a:fd:d8:61:e7:e4:eb:58:65:1e:ea:5a:4d:
+    4c:73:87:32:e0:91:a3:92:56:2e:a7:bc:1e:32:30:
+    43:f5:fd:db:05:5a:08:b2:25:15:5f:ac:4d:71:82:
+    2b:d0:87:b4:01" | tr -d ": \n" | xxd -p -r | base64 | tr +/ -_ | tr -d "=\n"
+tfJaLrzXILUg1U3N1KV8yJr92GHn5OtYZR7qWk1Mc4cy4JGjklYup7weMjBD9f3bBVoIsiUVX6xNcYIr0Ie0AQ
+```
+
+<note>
+<p>
+Note that the leading 00 byte has been omitted; that's an artifact of the ASN.1 encoding.
+</p>
+</note>
+
+By leveraging the `tr` command properly, the modulus field has been encoded into a Base64URL string that we can use in our `jwks.json` file.
+## Populating the jwks.json file
+
+In the previous steps, we gathered the following necessary information:
+
+1) An RSA key-pair
+2) The modulus of the RSA public key in Base64URL format
+3) The exponent of the RSA public key in Base64URL format
+
+With these at hand, we can now populate the [jwks.json](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets) file of our ktor project. 
+Simply specify the `e` and `n` values with the Base64URL encoded values we produced earlier respectively, specify a key id (in this case, we are using the `kid` from
+the sample project but you can use any string you like) and the `kty` attribute as `RSA`:
+
+```json
+{
+  "keys": [
+    {
+      "kty": "RSA",
+      "e": "AQAB",
+      "kid": "6f8856ed-9189-488f-9011-0ff4b6c08edc",
+      "n":"tfJaLrzXILUg1U3N1KV8yJr92GHn5OtYZR7qWk1Mc4cy4JGjklYup7weMjBD9f3bBVoIsiUVX6xNcYIr0Ie0AQ"
+    }
+  ]
+}
+```
+
+All that's left now, is to specify your private key so that your Ktor project can use it for authentication.
+
+## Defining the private key
+
+With your public key information set up, the last step is to provide your Ktor project with access
+to your private key.
+
+Assuming that you have extracted your private key (that you generated at the beginning in your `.pk8` file) into an environment variable on your system, called `jwt_pk` in this case, your `resources/application.conf` file's jwt section
+should look similar to:
+
+```
+jwt {
+  privateKey = ${jwt_pk}
+  issuer = "http://0.0.0.0:8080/"
+  audience = "http://0.0.0.0:8080/login"
+  realm = "MyProject"
+}
+```
+
+<tip>
+<p>
+Your private key is considered sensitive information and should not be stored directly in code. You should
+consider using environment variables for sensitive data.
```

Comment:
I suggest to change this to:

consider using environment variables or a secret store for sensitive data.

Reply:
Done!